crates/wasmtime/src/component/types.rs

```diff
@@ -66,23 +90,188 @@ impl<T: fmt::Debug> fmt::Debug for Handle<T> {
+            .fields
+            .iter()
+            .map(|f| (&f.name, &f.ty))
+            .collect::<HashMap<&String, &InterfaceType>>();
+        a.fields.iter().all(|a_field| {
+            let Some(&&b_field_ty) = b_fields.get(&a_field.name) else {
+                return false;
+            };
+            self.interface_types_equal(a_field.ty, b_field_ty)
+        })
+    }
+
+    fn variants_equal(&self, v1: TypeVariantIndex, v2: TypeVariantIndex) -> bool {
+        let a = &self.a_types[v1];
+        let b = &self.b_types[v2];
+        if a.cases.len() != b.cases.len() {
+            return false;
+        }
+        let b_cases = b
+            .cases
+            .iter()
+            .map(|f| (&f.name, &f.ty))
+            .collect::<HashMap<&String, &Option<InterfaceType>>>();
+        a.cases.iter().all(|a_case| {
+            let Some(&&b_case_ty) = b_cases.get(&a_case.name) else {
+                return false;
+            };
+            match (a_case.ty, b_case_ty) {
+                (Some(a_case_ty), Some(b_case_ty)) => {
+                    self.interface_types_equal(a_case_ty, b_case_ty)
+                }
+                (None, None) => true,
+                _ => false,
+            }
+        })
+    }
+
+    fn results_equal(&self, r1: TypeResultIndex, r2: TypeResultIndex) -> bool {
+        let a = &self.a_types[r1];
+        let b = &self.b_types[r2];
+        let oks = match (a.ok, b.ok) {
+            (Some(ok1), Some(ok2)) => self.interface_types_equal(ok1, ok2),
+            (None, None) => true,
+            _ => false,
+        };
+        if !oks {
+            return false;
+        }
+        match (a.err, b.err) {
+            (Some(err1), Some(err2)) => self.interface_types_equal(err1, err2),
+            (None, None) => true,
+            _ => false,
+        }
+    }
+
+    fn options_equal(&self, o1: TypeOptionIndex, o2: TypeOptionIndex) -> bool {
+        let a = &self.a_types[o1];
+        let b = &self.b_types[o2];
+        self.interface_types_equal(a.ty, b.ty)
+    }
+
+    fn enums_equal(&self, e1: TypeEnumIndex, e2: TypeEnumIndex) -> bool {
+        let a = &self.a_types[e1];
+        let b = &self.b_types[e2];
+        if a.names.len() != b.names.len() {
+            return false;
+        }
+
+        let b_names = b.names.iter().collect::<HashSet<&String>>();
+        a.names.iter().all(|a_name| b_names.contains(a_name))
+    }
+
+    fn tuples_equal(&self, t1: TypeTupleIndex, t2: TypeTupleIndex) -> bool {
+        let a = &self.a_types[t1];
+        let b = &self.b_types[t2];
+        if a.types.len() != b.types.len() {
+            return false;
+        }
+        a.types
+            .iter()
+            .zip(b.types.iter())
+            .all(|(&a, &b)| self.interface_types_equal(a, b))
+    }
+
+    fn flags_equal(&self, f1: TypeFlagsIndex, f2: TypeFlagsIndex) -> bool {
+        let a = &self.a_types[f1];
+        let b = &self.b_types[f2];
+        a.names == b.names
```

Comment:
One thing I was unsure of: are flags types with names in different orders the same type?

Reply:
More-or-less yes it's ok to reorder but for the component model MVP we're going to require exact equivalence